File: src/ados/sdk/drivers/camera.py
"""Camera driver base class.

A camera driver pumps frames from a physical or networked imaging
device into the agent's frame bus. Visible, thermal, depth, and
multi-spectral devices all share this interface.

The peripheral manager owns discovery and arbitration. A driver's job
is to answer :meth:`CameraDriver.discover` honestly, open sessions on
demand, and yield :class:`FrameBuffer` instances until the session
closes.

``CameraSession`` is whatever opaque state the driver needs (file
descriptors, decoder pipelines, vendor SDK handles). The host treats
it as a token and only passes it back to the driver's lifecycle and
streaming methods.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any, AsyncIterator


@dataclass(frozen=True)
class CameraCandidate:
    """A device a driver claims it can open.

    ``device_id`` should be stable across reboots when the underlying bus
    can produce a stable identifier (USB serial, fixed CSI lane, RTSP URL).
    Where stability is impossible (raw v4l2 ``/dev/videoN`` indexing) the
    driver should still produce a deterministic id within a single boot.
    """

    driver_id: str
    device_id: str
    label: str
    bus: str
    vid_pid: tuple[int, int] | None = None
    metadata: dict[str, Any] | None = None


@dataclass(frozen=True)
class CameraCapabilities:
    """Static capabilities a session reports after :meth:`CameraDriver.open`."""

    radiometric: bool
    bit_depth: int
    width: int
    height: int
    fps: float
    pixel_format: str
    streaming_protocol: str
    color_spaces: list[str]
    has_audio: bool = False


@dataclass(frozen=True)
class FrameBuffer:
    """One frame yielded by a camera driver.

    ``data`` is a ``memoryview`` so downstream consumers can avoid copying
    when feasible. ``radiometric_k`` carries the per-pixel temperature
    reconstruction matrix when the sensor is radiometric. Drivers may
    attach driver-specific fields under ``metadata``.
    """

    timestamp_ns: int
    sequence: int
    width: int
    height: int
    pixel_format: str
    data: memoryview
    radiometric_k: memoryview | None = None
    metadata: dict[str, Any] | None = None


class CameraSession:
    """Opaque per-open state held by a driver.

    Concrete drivers subclass this with whatever fields they need. The
    base class is intentionally empty so the host only treats sessions
    as tokens.
    """


class CameraDriver(ABC):
    """Abstract base for camera drivers.

    Implementations register themselves with ``peripheral_manager`` from
    a plugin's ``on_start`` hook. The host then calls :meth:`discover`,
    arbitrates among candidates, and routes selected devices through
    :meth:`open`.
    """

    @abstractmethod
    async def discover(self) -> list[CameraCandidate]:
        """Scan for devices this driver can open.

        Called on agent boot and on hotplug events. Returns the list of
        candidate devices the driver is willing to claim. An empty list
        means the driver found nothing it recognises and is fine.
        """

    @abstractmethod
    async def open(
        self, candidate: CameraCandidate, config: dict[str, Any]
    ) -> CameraSession:
        """Open a session against a device.

        Raises :class:`ados.sdk.drivers.errors.DriverError` (or a subclass)
        on failure. ``config`` carries driver-specific options validated
        against the plugin's ``config-schema.json``.
        """

    @abstractmethod
    async def close(self, session: CameraSession) -> None:
        """Release resources held by a session."""

    @abstractmethod
    def capabilities(self, session: CameraSession) -> CameraCapabilities:
        """Return the static capabilities of an open session."""

    @abstractmethod
    async def frame_iterator(
        self, session: CameraSession
    ) -> AsyncIterator[FrameBuffer]:
        """Yield frames until the session closes or the iterator is cancelled."""

    @abstractmethod
    async def set_param(
        self, session: CameraSession, param: str, value: Any
    ) -> None:
        """Set a runtime parameter (gain, exposure, palette, shutter, etc.).

        Drivers should raise ``ValueError`` for unknown parameter names so
        the GCS panel can surface the rejection cleanly.
        """

File: crates/ados-sdk/src/drivers/camera.rs
//! Camera driver trait.
//!
//! Ports `ados.sdk.drivers.camera`. A camera driver pumps frames from a
//! physical or networked imaging device. Visible, thermal, depth, and
//! multi-spectral devices share this interface. The host owns discovery and
//! arbitration; the driver answers [`CameraDriver::discover`], opens sessions,
//! and yields [`FrameBuffer`] frames until the session closes.

use std::collections::BTreeMap;

use async_trait::async_trait;
use rmpv::Value;

use super::{Bus, DriverResult, SampleStream};

/// A device a driver claims it can open. `device_id` should be stable across
/// reboots when the bus exposes a stable identifier (USB serial, fixed CSI
/// lane, RTSP URL); otherwise it must at least be deterministic within one boot.
#[derive(Debug, Clone, PartialEq)]
pub struct CameraCandidate {
    pub driver_id: String,
    pub device_id: String,
    pub label: String,
    pub bus: Bus,
    pub vid_pid: Option<(u16, u16)>,
    pub metadata: BTreeMap<String, Value>,
}

/// Static capabilities a session reports after [`CameraDriver::open`].
#[derive(Debug, Clone, PartialEq)]
pub struct CameraCapabilities {
    pub radiometric: bool,
    pub bit_depth: u32,
    pub width: u32,
    pub height: u32,
    pub fps: f32,
    pub pixel_format: String,
    pub streaming_protocol: String,
    pub color_spaces: Vec<String>,
    pub has_audio: bool,
}

/// One frame yielded by a camera driver. `radiometric_k` carries the per-pixel
/// temperature reconstruction matrix when the sensor is radiometric.
#[derive(Debug, Clone, PartialEq)]
pub struct FrameBuffer {
    pub timestamp_ns: i64,
    pub sequence: u64,
    pub width: u32,
    pub height: u32,
    pub pixel_format: String,
    pub data: Vec<u8>,
    pub radiometric_k: Option<Vec<u8>>,
    pub metadata: BTreeMap<String, Value>,
}

/// Abstract camera driver. Implementations register with the peripheral
/// manager from `on_start`. The host then calls [`CameraDriver::discover`],
/// arbitrates among candidates, and routes selected devices through
/// [`CameraDriver::open`].
///
/// `Session` is the driver's opaque per-open state (file descriptors, decoder
/// pipelines, vendor SDK handles). The host treats it as a token and only
/// passes it back to the driver's lifecycle and streaming methods.
#[async_trait]
pub trait CameraDriver: Send + Sync {
    /// Opaque per-open state. The Rust analogue of the Python `CameraSession`
    /// base class.
    type Session: Send + Sync;

    /// Scan for devices this driver can open. Called on boot and on hotplug. An
    /// empty list means the driver recognised nothing and is fine.
    async fn discover(&self) -> DriverResult<Vec<CameraCandidate>>;

    /// Open a session against a device. `config` carries driver-specific
    /// options validated against the plugin's `config-schema.json`.
    async fn open(
        &self,
        candidate: &CameraCandidate,
        config: &BTreeMap<String, Value>,
    ) -> DriverResult<Self::Session>;

    /// Release resources held by a session.
    async fn close(&self, session: Self::Session) -> DriverResult<()>;

    /// Return the static capabilities of an open session.
    fn capabilities(&self, session: &Self::Session) -> CameraCapabilities;

    /// Yield frames until the session closes or the stream is dropped. The
    /// Rust analogue of the Python `frame_iterator` `AsyncIterator`.
    async fn frame_iterator(
        &self,
        session: &Self::Session,
    ) -> DriverResult<SampleStream<FrameBuffer>>;

    /// Set a runtime parameter (gain, exposure, palette, shutter). Return
    /// [`super::DriverError::InvalidParam`] for unknown parameter names so the
    /// GCS panel can surface the rejection.
    async fn set_param(
        &self,
        session: &Self::Session,
        param: &str,
        value: Value,
    ) -> DriverResult<()>;
}
